graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.cluster.lock;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.mongodb.MongoCommandException;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.Indexes;
+import com.mongodb.client.model.ReturnDocument;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.plugin.system.NodeId;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Updates.currentDate;
+import static org.graylog2.cluster.lock.Lock.FIELD_LOCKED_BY;
+import static org.graylog2.cluster.lock.Lock.FIELD_RESOURCE_NAME;
+import static org.graylog2.cluster.lock.Lock.FIELD_UPDATED_AT;
+
+/**
+ * Lock service implementation using MongoDB to maintain locks.
+ * Unless a lock is kept alive by periodically re-requesting the same lock, it will eventually expire.
+ * Lock expiry is handled by MongoDB internally. We set lock expiry to 60 seconds, but in practice it may take up to ~2
+ * minutes until a lock really expires.
+ */
+@Singleton
+public class MongoLockService implements LockService {
+
+    public static final String COLLECTION_NAME = "cluster_locks";
+    public static final java.time.Duration MIN_LOCK_TTL = Duration.ofSeconds(60);
+
+    private final NodeId nodeId;
+    private final MongoCollection<Document> collection;
+
+    @Inject
+    public MongoLockService(NodeId nodeId,
+                            MongoConnection mongoConnection,
+                            @Named("leader_election_lock_ttl") Duration leaderElectionLockTTL) {
```

Comment:
We currently use the `LockService` only for the leader election. But I think we might use the lock service for other things as well in the future. (e.g. job scheduler)
How about renaming this config option to something more generic? (e.g. `lock_service_lock_ttl`)

Reply:
Addressed in https://github.com/Graylog2/graylog2-server/pull/11720